src/main/java/org/zalando/nakadi/service/subscription/state/PartitionData.java

```diff
@@ -52,25 +53,58 @@
         this.commitOffset = commitOffset;
         this.sentOffset = commitOffset;
         this.lastSendMillis = currentTime;
+        this.batchWindowStartTimestamp = 0L;
     }
 
     @Nullable
-    List<ConsumedEvent> takeEventsToStream(final long currentTimeMillis, final int batchSize,
-                                           final long batchTimeoutMillis, final boolean streamTimeoutReached) {
+    List<ConsumedEvent> takeEventsToStream(final long currentTimeMillis, final long batchTimespanMillis,
+                                           final int batchSize, final long batchTimeoutMillis,
+                                           final boolean streamTimeoutReached) {
         final boolean countReached = (nakadiEvents.size() >= batchSize) && batchSize > 0;
         final boolean timeReached = (currentTimeMillis - lastSendMillis) >= batchTimeoutMillis;
-        if (countReached || timeReached) {
+        final long lastRecordTimestamp = lastRecordTimestamp();
+
+        if (batchWindowStartTimestamp == 0 && !nakadiEvents.isEmpty()) {
+            batchWindowStartTimestamp = nakadiEvents.get(0).getTimestamp();
+        }
+
+        if (batchTimespanMillis > 0 && lastRecordTimestamp > 0
+                && lastRecordTimestamp >= batchWindowStartTimestamp + batchTimespanMillis) {
+            final long batchWindowEndTimestamp = batchWindowStartTimestamp + batchTimespanMillis;
+            batchWindowStartTimestamp = batchWindowEndTimestamp;
             lastSendMillis = currentTimeMillis;
+            return extractTimespan(batchWindowEndTimestamp);
+        } else if (countReached || timeReached) {
+            lastSendMillis = currentTimeMillis;
+            batchWindowStartTimestamp = lastSendMillis;
             return extract(batchSize);
         } else if (streamTimeoutReached) {
             lastSendMillis = currentTimeMillis;
+            batchWindowStartTimestamp = lastSendMillis;
             final List<ConsumedEvent> extractedEvents = extract(batchSize);
             return extractedEvents.isEmpty() ? null : extractedEvents;
         } else {
             return null;
         }
     }
 
+    private long lastRecordTimestamp() {
+        if (nakadiEvents.size() > 0) {
+            return nakadiEvents.get(nakadiEvents.size() - 1).getTimestamp();
+        } else {
+            return 0;
+        }
+    }
+
+    private List<ConsumedEvent> extractTimespan(final long batchWindowEndTimestamp) {
+        int count = 1;
+        while (count < nakadiEvents.size() && nakadiEvents.get(count).getTimestamp() < batchWindowEndTimestamp) {
```

Comment:
I would really like to optimize, but having another sorted data structure would cost more than just having iterative approach

Reply:
I understand your point. I thought about it and I considered it to not be a problem because this code is only invoked when the flush condition is met, so it's not run on every call to `takeEventsToStream`. There we always take a look at the last event that was loaded from Kafka `O(1)` and then once it matches the criteria, then we run a `O(n)` scan to find the element. I could even optimize it by doing some binary search, as I would expect `getTimestamp` to be more or less sorted (there could be some anomalies, but that would not a problem, as overall it's almost guaranteed to be sorted).

I could implement the binary search. But it would not improve the complexity, as copying the data takes `O(n)` anyway.

Reply:
I can also not rely on existing `extract` method and implement my own, that copies data as it checks it, avoiding going over the data twice.

Reply:
e8e2069b